### ISSUE TITLE
Why is inference slower when using fit_with_cache for a single sample?

### ISSUE THREAD

**Issue by lindage** [PriorLabs/TabPFN#904]
### Describe the bug

tabpfn 7.1.1  

```
import time
import numpy as np
from tabpfn import TabPFNClassifier

X_train = np.random.randn(442, 10)
y_train = np.random.randint(0, 2, size=(442,)) 

X_test = np.random.randn(1, 10)
X_test2 = np.random.randn(1, 10)

for fit_mode in ['low_memory', 'fit_preprocessors', 'fit_with_cache']:
    print(f"\nFit mode: {fit_mode}")

    classifier = TabPFNClassifier(
        random_state=42, 
        fit_mode=fit_mode,
        model_path='./tabpfn-v2.5-classifier-v2.5_default.ckpt', 
    )

    start = time.time()
    classifier.fit(X_train, y_train.ravel())
    end = time.time()
    print(f"Time elapsed fit:          {end - start}")

    start = time.time()
    predictions = classifier.predict(X_test)
    end = time.time()
    print(f"Time elapsed predict 1:    {end - start}")

    start = time.time()
    predictions = classifier.predict(X_test2)
    end = time.time()
    print(f"Time elapsed predict 2:    {end - start}")
```

#############
Fit mode: low_memory
Time elapsed fit:          5.017396926879883
Time elapsed predict 1:    0.2986717224121094
Time elapsed predict 2:    0.30551624298095703

Fit mode: fit_preprocessors
Time elapsed fit:          4.7202184200286865
Time elapsed predict 1:    0.4782838821411133
Time elapsed predict 2:    0.22461295127868652

Fit mode: fit_with_cache
Time elapsed fit:          25.51187300682068
Time elapsed predict 1:    0.5878572463989258
Time elapsed predict 2:    0.7825231552124023

### Steps/Code to Reproduce

See above

### Expected Results

See above

https://github.com/PriorLabs/TabPFN/issues/457    It's different from this one (fit_with_cache is fastest)

### Actual Results

See above

### Versions

```shell
Collecting system and dependency information...
PyTorch version: 2.9.1+cu128
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Alibaba Group Enterprise Linux Server 7.2 (Paladin) (x86_64)
GCC version: (GCC) 10.2.1 20200825 (Alibaba 10.2.1-3 2.17)
Clang version: Could not collect
CMake version: version 3.31.2
Libc version: glibc-2.32

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.10.134-13.al8.x86_64-x86_64-with-glibc2.32
Is CUDA available: True
CUDA runtime version: 12.8.61
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA A800-SXM4-80GB
GPU 1: NVIDIA A800-SXM4-80GB
GPU 2: NVIDIA A800-SXM4-80GB
GPU 3: NVIDIA A800-SXM4-80GB

Nvidia driver version: 470.161.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                128
On-line CPU(s) list:   0-127
Thread(s) per core:    2
Core(s) per socket:    32
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 106
Model name:            Intel(R) Xeon(R) Platinum 8369B CPU @ 2.90GHz
Stepping:              6
CPU MHz:               3500.002
CPU max MHz:           3500.0000
CPU min MHz:           800.0000
BogoMIPS:              5800.00
Virtualization:        VT-x
L1d cache:             48K
L1i cache:             32K
L2 cache:              1280K
L3 cache:              49152K
NUMA node0 CPU(s):     0-31,64-95
NUMA node1 CPU(s):     32-63,96-127
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid fsrm md_clear pconfig flush_l1d arch_capabilities

Dependency Versions:
--------------------
tabpfn: 7.1.1
torch: 2.9.1
numpy: 1.26.4
scipy: 1.15.3
pandas: 2.0.1
scikit-learn: 1.3.2
typing_extensions: 4.15.0
einops: 0.8.2
huggingface-hub: 0.36.2
```

**Comment by adrian-prior** [comment]
Thanks for the detailed report and repro script! 🙏

The short version: **this is expected behavior, and `fit_with_cache` is designed to trade a slower `fit` for faster repeated `predict`s.** It precomputes and caches the transformer's representation of the training context at fit time, so each subsequent inference can reuse it instead of recomputing it. That makes the one `fit` call more expensive, but every `predict` after it cheaper. If you only ever call `predict` once, the caching overhead isn't amortized and `fit_with_cache` is the wrong choice — `fit_preprocessors` (the default) is what you want there.

I also wanted to flag that your numbers are on **tabpfn 7.1.1 with the v2.5 checkpoint**. On the latest release (**tabpfn 8.0.8, the TabPFN-3 model**) I couldn't reproduce the *slower predict* you saw — there, `fit_with_cache` actually makes `predict` ~40% faster, as intended.

To make the comparison robust I ran your benchmark over **10 repetitions (plus 2 warmup)** with proper CUDA synchronization, reporting mean ± standard error. On an **NVIDIA A100-40GB**, tabpfn 8.0.8, default TabPFN-3 classifier:

| fit_mode | fit | predict 1 | predict 2 |
|---|---|---|---|
| `low_memory` | 636.3 ± 26.4 ms | 658.8 ± 2.8 ms | 659.4 ± 4.1 ms |
| `fit_preprocessors` | 626.6 ± 13.5 ms | 646.8 ± 0.9 ms | 646.4 ± 1.1 ms |
| `fit_with_cache` | 1577.9 ± 27.6 ms | **391.6 ± 14.6 ms** | **377.2 ± 2.5 ms** |

So `fit_with_cache` ≈ 2.5× slower `fit`, but ≈ 40% faster `predict` — the intended trade-off. **Rule of thumb:** use `fit_with_cache` when you fit once and predict many times; use the default `fit_preprocessors` for one-shot fit→predict.

A couple of notes:
- A single `time.time()` around a `predict` on GPU can be misleading because CUDA kernels run asynchronously — the call can return before the GPU finishes. Adding `torch.cuda.synchronize()` around the timed region (as below) gives accurate wall-clock numbers.
- These are GPU numbers; on CPU the absolute times are much larger and the relative benefit of caching can differ. If your workload is CPU-bound and latency-sensitive, it's worth benchmarking both modes on your own hardware.

Benchmark script (your original, adapted to the latest model, multiple reps, and synchronized timing):

```python
import time
import statistics
import numpy as np
import torch
from tabpfn import TabPFNClassifier

N_REP = 10
WARMUP = 2
DEVICE = "cuda"  # or "cpu"
MODES = ["low_memory", "fit_preprocessors", "fit_with_cache"]


def sync() -> None:
    if torch.cuda.is_available():
        torch.cuda.synchronize()


def stats(xs):
    m = statistics.mean(xs)
    se = statistics.stdev(xs) / len(xs) ** 0.5 if len(xs) > 1 else 0.0
    return m, se


rng = np.random.default_rng(42)
X_train = rng.standard_normal((442, 10))
y_train = rng.integers(0, 2, size=442)
X_test = rng.standard_normal((1, 10))
X_test2 = rng.standard_normal((1, 10))

for mode in MODES:
    fit_t, p1_t, p2_t = [], [], []
    for rep in range(WARMUP + N_REP):
        clf = TabPFNClassifier(random_state=42, fit_mode=mode, device=DEVICE)
        sync(); s = time.perf_counter(); clf.fit(X_train, y_train); sync(); t_fit = time.perf_counter() - s
        sync(); s = time.perf_counter(); clf.predict(X_test); sync(); t_p1 = time.perf_counter() - s
        sync(); s = time.perf_counter(); clf.predict(X_test2); sync(); t_p2 = time.perf_counter() - s
        if rep >= WARMUP:
            fit_t.append(t_fit); p1_t.append(t_p1); p2_t.append(t_p2)
    print(f"Fit mode: {mode}")
    for name, xs in [("fit", fit_t), ("predict 1", p1_t), ("predict 2", p2_t)]:
        m, se = stats(xs)
        print(f"  {name:<11}: {m * 1000:9.2f} ms +/- {se * 1000:6.2f} ms  (n={len(xs)})")
    print()
```

If you're on an older version, I'd recommend upgrading to the latest `tabpfn` and using the default TabPFN-3 checkpoint — and if single-`predict` latency is your priority, stick with `fit_preprocessors`. Hope this helps!
